### PR TITLE
feat: add spawn and try_current for madsim-tokio

### PR DESCRIPTION
**Pull request by KveinAxel**
to support madsim-rs/sqlx, we add some `spawn`, `try_current` and some other dummy interface.

### CODE REVIEW

File: madsim/src/sim/runtime/mod.rs [hunk 1]
```diff
@@ -226,6 +231,40 @@ impl Handle {
...
+    }
+
+    /// spawn a task
+    pub fn spawn<F>(&self, future: F) -> JoinHandle<F::Output>
+    where
+        F: Future + Send + 'static,
+        F::Output: Send + 'static,
+    {
+        self.create_node()
+            .name("spawn a task")
```
Comment (wangrunji0408) [inline madsim/src/sim/runtime/mod.rs:250]: We shouldn't create a new node here. The task should be spawned within the current node. Please use `Spawner::current()` instead.

File: madsim/src/sim/runtime/mod.rs [hunk 2]
```diff
@@ -226,6 +231,40 @@ impl Handle {
...
+    #[deprecated(
+        since = "0.3.0",
+        note = "blocking function is not allowed in simulation"
+    )]
+    pub fn spawn_blocking<F, R>(&self, _f: F) -> JoinHandle<R>
+    where
+        F: FnOnce() -> R + Send + 'static,
+        R: Send + 'static,
+    {
+        unimplemented!()
```
Comment (wangrunji0408) [inline madsim/src/sim/runtime/mod.rs:265]:
```suggestion
        unimplemented!("blocking function is not allowed in simulation")
```

**Review by wangrunji0408 — APPROVED** [review]
Rest LGTM! Thanks

File: madsim-tokio/src/sim/runtime.rs [hunk 3]
```diff
@@ -71,6 +72,10 @@ impl Runtime {
         handle
     }
 
+    pub fn block_on<F: Future>(&self, _future: F) -> F::Output {
+        unimplemented!();
+    }
```
[the remark this answers is not in the document]
Comment (KveinAxel) [reply madsim-tokio/src/sim/runtime.rs:77]: It's in the https://github.com/madsim-rs/sqlx/blob/d4bff0391aa4c7300f9afbcf8ec7b0d353adcfa4/sqlx-core/src/rt/mod.rs#L123 and which is just used for test, we can add a dummy interface here, or remove all the `test_block_on` tree in sqlx

File: madsim/src/sim/net/unix/stream.rs [hunk 4]
```diff
@@ -43,6 +44,10 @@ impl UnixStream {
     {
         todo!();
     }
+
+    pub fn try_read_buf<B: BufMut>(&mut self, buf: &mut B) -> io::Result<usize> {
+        unimplemented!();
+    }
```
[the remark this answers is not in the document]
Comment (KveinAxel) [reply madsim/src/sim/net/unix/stream.rs:50]: This just used to pass the sqlx's compile, so maybe we can just leave a  unimplemented here? So far sqlx hasn't called this on `TcpStream` yet.

File: madsim/src/sim/runtime/mod.rs (hunk 1 above)
Comment (KveinAxel) [reply madsim/src/sim/runtime/mod.rs:250]: I have tried `Spawner::current` before, but in some cases, like dropping some `Future` and calling a sync `drop` to spawn an async task which means there are no tasks in the runtime. The drop site is out the scope of https://github.com/madsim-rs/madsim/blob/ca933529dca42bcc5203a98268365fd778a05676/madsim/src/sim/task/mod.rs#L270

File: madsim/src/sim/net/tcp/stream.rs [hunk 5]
```diff
@@ -129,7 +144,8 @@ impl AsyncRead for TcpStream {
             return Poll::Ready(Ok(()));
         }
         // otherwise wait on channel
-        match self.rx.poll_next_unpin(cx) {
+        let poll_res = { self.rx.poll_next_unpin(cx) };
+        match poll_res {
```
[the remark this answers is not in the document]
Comment (KveinAxel) [reply madsim/src/sim/net/tcp/stream.rs:148]: Maybe some version of checker required this, I'll rollback this code.

File: madsim/src/sim/runtime/mod.rs (hunk 1 above)
Comment (wangrunji0408) [reply madsim/src/sim/runtime/mod.rs:250]: I see. In this case, I think we should enter task context before dropping the task. And here we should still use `Spawner::current()`.

**Pull request merged**